protocol/handlerelayinvs/handle_relay_invs.go

```diff
@@ -0,0 +1,218 @@
+package handlerelayinvs
+
+import (
+	"github.com/kaspanet/kaspad/blockdag"
+	"github.com/kaspanet/kaspad/netadapter"
+	"github.com/kaspanet/kaspad/protocol/blocklogger"
+	peerpkg "github.com/kaspanet/kaspad/protocol/peer"
+	"github.com/kaspanet/kaspad/util"
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/kaspanet/kaspad/wire"
+	"github.com/pkg/errors"
+	"time"
+)
+
+// HandleRelayInvs listens to wire.MsgInvRelayBlock messages, requests their corresponding blocks if they
+// are missing, adds them to the DAG and propagates them to the rest of the network.
+func HandleRelayInvs(msgChan <-chan wire.Message, peer *peerpkg.Peer, netAdapter *netadapter.NetAdapter, router *netadapter.Router,
+	dag *blockdag.BlockDAG) error {
+
+	invsQueue := make([]*wire.MsgInvRelayBlock, 0)
+	for {
+		inv, shouldStop, err := readInv(msgChan, &invsQueue)
+		if err != nil {
+			return err
+		}
+		if shouldStop {
+			return nil
+		}
+
+		if dag.IsKnownBlock(inv.Hash) {
+			if dag.IsKnownInvalid(inv.Hash) {
+				return errors.Errorf("sent inv of an invalid block %s",
+					inv.Hash)
+			}
+			return nil
+		}
+
+		requestQueue := newHashesQueueSet()
+		requestQueue.enqueueIfNotExists(inv.Hash)
+
+		for requestQueue.len() > 0 {
+			shouldStop, err := requestBlocks(netAdapter, router, peer, msgChan, dag, &invsQueue,
+				requestQueue)
+			if err != nil {
+				return err
+			}
+			if shouldStop {
+				return nil
+			}
+		}
+	}
+}
+
+func readInv(msgChan <-chan wire.Message,
+	invsQueue *[]*wire.MsgInvRelayBlock) (inv *wire.MsgInvRelayBlock, shouldStop bool, err error) {
+
+	if len(*invsQueue) > 0 {
+		inv, *invsQueue = (*invsQueue)[0], (*invsQueue)[1:]
+		return inv, false, nil
+	}
+
+	msg, isOpen := <-msgChan
+	if !isOpen {
+		return nil, true, nil
+	}
+
+	inv, ok := msg.(*wire.MsgInvRelayBlock)
+	if ok {
+		return inv, false, nil
```

Comment:
I would prefer if you'd handle the error case here and the "good" case at the bottom of the function.